cmd/webhook-manager/app/util.go

```diff
@@ -17,93 +17,96 @@ limitations under the License.
+	"bytes"
 	"context"
 	"crypto/tls"
 	"crypto/x509"
-	"regexp"
-	"strings"
+	"fmt"
+	"time"
 
 	v1 "k8s.io/api/admissionregistration/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/rest"
 	"k8s.io/klog"
 
 	"volcano.sh/apis/pkg/client/clientset/versioned"
 	"volcano.sh/volcano/cmd/webhook-manager/app/options"
-	"volcano.sh/volcano/pkg/webhooks/router"
 )
 
-func registerWebhookConfig(kubeClient *kubernetes.Clientset, config *options.Config, service *router.AdmissionService, caBundle []byte) {
-	sideEffect := v1.SideEffectClassNoneOnDryRun
-	reviewVersions := []string{"v1"}
-	webhookLabelSelector := &metav1.LabelSelector{}
-	clientConfig := v1.WebhookClientConfig{
-		CABundle: caBundle,
-	}
-	if config.WebhookURL != "" {
-		url := config.WebhookURL + service.Path
-		clientConfig.URL = &url
-		klog.Infof("The URL of webhook manager is <%s>.", url)
+var (
+	validatingWebhooksName = []string{
+		"volcano-admission-service-jobs-validate",
+		"volcano-admission-service-pods-validate",
+		"volcano-admission-service-queues-validate",
 	}
-	if config.WebhookName != "" && config.WebhookNamespace != "" {
-		clientConfig.Service = &v1.ServiceReference{
-			Name:      config.WebhookName,
-			Namespace: config.WebhookNamespace,
-			Path:      &service.Path,
-		}
-		klog.Infof("The service of webhook manager is <%s/%s/%s>.",
-			config.WebhookName, config.WebhookNamespace, service.Path)
+	mutatingWebhooksName = []string{
+		"volcano-admission-service-pods-mutate",
+		"volcano-admission-service-queues-mutate",
+		"volcano-admission-service-podgroups-mutate",
+		"volcano-admission-service-jobs-mutate",
 	}
-	if config.IgnoredNamespaces != "" {
-		ignoredNamespaces := strings.Split(strings.TrimSpace(config.IgnoredNamespaces), ",")
-		klog.Infof("The ignored namespaces list of webhook manager is <%v>.",
-			ignoredNamespaces)
-		webhookLabelSelector = &metav1.LabelSelector{
-			MatchExpressions: []metav1.LabelSelectorRequirement{
-				{
-					Values:   ignoredNamespaces,
-					Operator: "NotIn",
-					Key:      "kubernetes.io/metadata.name",
-				},
-			},
-		}
-	}
-	if service.MutatingConfig != nil {
-		for i := range service.MutatingConfig.Webhooks {
-			service.MutatingConfig.Webhooks[i].SideEffects = &sideEffect
-			service.MutatingConfig.Webhooks[i].AdmissionReviewVersions = reviewVersions
-			service.MutatingConfig.Webhooks[i].ClientConfig = clientConfig
-			service.MutatingConfig.Webhooks[i].NamespaceSelector = webhookLabelSelector
-		}
+)
 
-		service.MutatingConfig.ObjectMeta.Name = webhookConfigName(config.WebhookName, service.Path)
+func addCaCertForWebhook(kubeClient *kubernetes.Clientset, caBundle []byte) error {
+	for _, mutatingWebhookName := range mutatingWebhooksName {
+		var mutatingWebhook *v1.MutatingWebhookConfiguration
+		if err := wait.PollInfinite(time.Second, func() (done bool, err error) {
```

Comment:
it's better to add timeout and exit

Reply:
How long should the timeout be set?

Reply:
@shinytang6
I added a 5 minutes timeout.